I2C: Enable intel's arch/drivers/boards to DTS

tbursztyka opened the pull request:
galileo, all the quark_se(_ss) devboard and the arduino_101(_sss)

Commit:
dts/i2c: Fixing copyright headers

Some yaml files were missing there copyright header.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Cleanup Kconfig

Move driver specific to dedicated file when relevant (i.e.: more than
1-2 options), use if/endif also.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Indent sources the same way everywhere

Using tabs only.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the DW i2c driver

Shared irq is not managed there as it is a very SoC specific issue.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in quark_x1000

This will generate the necessary configuration for the i2c controller on
quark_x1000.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/galileo: Switch to DTS for generating i2c settings

Thus dropping SoC based generated information.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Rename generic configure option in DW drivers

And adapt SoC's information relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI i2c driver

Used in some Intel quark SoC.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in intel curie

This will generate the necessary configuration for the 2 i2c controllers
on quark_se_c1000.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Enable i2c through DTS for quark_se

And adapt relevant boards accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in quark_d2000

This will generate the necessary configuration for the i2c controller
on quark_d2000.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Enable i2c through DTS for quark_d2000

And adapt the relevant board accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI SS i2c driver

Used in Intel quark_se's ARC core.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in Quark SE SS C1000

This will generate the necessary configuration for the 2 i2c controllers
on quark_se_c1000_ss.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Disable some QMSI SS option when HAS_DTS_I2C is enabled

These options will be generated through DTS.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Enable i2c through DTS for quark_se SS

And adapt relevant boards accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

galak commented on dts/bindings/i2c/synopsys,dw-i2c.yaml:
```diff
@@ -0,0 +1,36 @@
+#
+---
+title: DesignWare SPI node description
+id: synopsys,dw-i2c
```
We should use/conform to the linux binding Documentation/devicetree/bindings/i2c/i2c-designware.txt (for things like compat, etc)

galak commented on dts/x86/quark_x1000.dtsi:
```diff
@@ -59,5 +59,16 @@
 		};
+
+		i2c0: i2c@0x90007000 {
+			compatible = "synopsys,dw-i2c";
```
Should have #address-cells & #size-cells since its a bus

galak commented on dts/x86/intel_curie.dtsi:
```diff
@@ -92,5 +92,23 @@
+			clock-frequency = <0x12>;
+			label = "I2C_0";
+
+			status = "disabled";
```
Should have #address-cells & #size-cells since its a bus

galak commented on dts/x86/intel_curie.dtsi:
```diff
@@ -92,5 +92,23 @@
+			reg = <0xb0002c00 0x400>;
+			clock-frequency = <0x12>;
+			label = "I2C_1";
+
```
Should have #address-cells & #size-cells since its a bus

galak commented on dts/x86/intel_quark_d2000.dtsi:
```diff
@@ -77,5 +77,14 @@
+			reg = <0xb0002800 0x400>;
+			clock-frequency = <0x12>;
+			label = "I2C_0";
+
```
Should have #address-cells & #size-cells since its a bus

galak reviewed (changes requested):
* Need addr/size cell in all dts
* we need to fix clock frequency / default_cfg:

Take a look at commit 54933b38339dcfa5efcfc3b47d35b197699b2eb9 for example

Commit:
drivers/i2c: Removing useless configuration attribute in DW driver

Probably a left over form legacy interrupt handling.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Use standard bitrate settings for DW driver

Applying the change to relevant arch/boards, either in their Kconfig or
the dts specific files.

Taking the opportunity in dw driver to rename the variable the same way
as they are everywhere else in the code (s/dev/dw and s/port/dev) in
init function.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Use standard bitrate settings for QMSI driver

Apply the change on relevant arch/boards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Use standard bitrate settings for QMSI SS driver

Apply the change on relevant arch/boards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

ydamigos commented on arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series:
```diff
@@ -24,7 +24,10 @@ endif # GPIO_STM32
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V2
+config I2C_STM32
```
We should move this in arch/arm/soc/st_stm32/common/Kconfig.defconfig.series
```
if I2C && (I2C_1 || I2C_2 || I2C_3)
config I2C_STM32
	def_bool y
endif # I2C
```
and leave here
```
if I2C_STM32
config I2C_STM32_V2
 	def_bool y
endif # I2C_STM32
```

ydamigos commented on arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series:
```diff
@@ -21,7 +21,10 @@ endif # GPIO_STM32
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V1
+config I2C_STM32
```
Same here

ydamigos commented on arch/arm/soc/st_stm32/stm32f3/Kconfig.defconfig.series:
```diff
@@ -21,7 +21,10 @@ endif # GPIO_STM32
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V2
+config I2C_STM32
```
Same here

ydamigos commented on arch/arm/soc/st_stm32/stm32f4/Kconfig.defconfig.series:
```diff
@@ -47,7 +47,10 @@ endif #USB
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V1
+config I2C_STM32
```
Same here

ydamigos commented on arch/arm/soc/st_stm32/stm32l4/Kconfig.defconfig.series:
```diff
@@ -15,7 +15,10 @@ config SOC_SERIES
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V2
+config I2C_STM32
```
Same here

ydamigos commented on dts/x86/quark_x1000.dtsi:
```diff
@@ -59,5 +59,17 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x90007000 0x400>;
+			interrupts = <18 IRQ_TYPE_LEVEL_LOW>;
```
Is IRQ_TYPE_LEVEL_LOW the IRQ priority?

ydamigos commented on arch/x86/soc/intel_quark/quark_x1000/Kconfig.defconfig.series:
```diff
@@ -70,12 +70,11 @@ config I2C_0
-	default "I2C_0"
 config I2C_0_DEFAULT_CFG
 	default 0x12
 config I2C_0_IRQ_PRI
```
You can get IRQ_PRI from DT.

ydamigos replied:
You can get IRQ_PRI from DT.

ydamigos commented on arch/x86/soc/intel_quark/quark_se/Kconfig.defconfig.series:
```diff
@@ -86,6 +86,7 @@ config I2C_1
 
 if I2C_1
+
 config I2C_1_IRQ_PRI
```
I am sure that I am missing something here. There is no interrupt property in intel,qmsi-i2c but there is IRQ_PRI here. 
Same for I2C_0.

ydamigos commented on arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series:
```diff
@@ -99,6 +99,7 @@ config I2C_QMSI
 config I2C_0
 	def_bool y
+
 config I2C_0_IRQ_PRI
```
I am sure that I am missing something here. There is no interrupt property in intel,qmsi-i2c but there is IRQ_PRI here.

tbursztyka replied:
Indeed, sounds like some old left-over.

tbursztyka replied:
Actuall, we both overlooked: it is there. Difference being that irq connection is done in init function.

ydamigos replied:
I was talking about intel,qmsi-i2c.dtsi file. Will you add support to the i2c_qmsi driver to get IRQ configuration from DT.

tbursztyka replied:
aah, ok. I tried that but interrupts can't take more than 2 arguments there. Adding a interrupt-priority attribute could work, but not nice.

ydamigos replied:
Did you try to change #interrupt-cells number in intc node?

tbursztyka replied:
Ah, that could help, thanks. I am new to dts, thus the trial & error path...

tbursztyka replied:
Does not work, because of the interrupt-parent, &intc sets the interrupt-cells to 2.

ydamigos replied:
I meant to change &intc node the interrupt-cells to 3 and update all the nodes using it (uart, i2c spi, ...).

tbursztyka replied:
oh, true. I need a break.

Will see that.

tbursztyka replied:
Does not work, seems like a bug in extract_dts_includes.py to me but I am not entirely sure. (this tool _really_ needs to be rewritten)

ydamigos replied:
I think you should modify also intel,ioapic.yaml from:
```
cell_string: IRQ

"#cells":
  - irq
  - sense
```
to:
```
cell_string: IRQ

"#cells":
  - irq
  - priority
  - sense
```
but I am not sure.

galak replied:
Does the intel,ioapic support priorities?  From what I can tell it doesn't seem to.  The binding for ioapic yaml is based on the linux binding if I remember correctly.

tbursztyka replied:
It does support priorities. I don't know why Linux does not expose this setting, could it be because the intel platforms supported there provide all the necessary information through PCI?

galak replied:
Looking at the driver in zephyr, it doesn't seem that we allow or have means to set the priority in the ioapic (but I could be missing something).

tbursztyka replied:
I am not really aware of the details in that part, but it seems that ioapic itself does not manage priorities directly, hardware wise. It seems to be handled in software through the ISR table.

andrewboie replied:
We *do* set priorities. But it is not done in the IOAPIC. Priorities are reflected by the vector assignment in the IDT. Based on the priority level given to IRQ_CONNECT(), gen_idt uses this to find an appropriate range in the vector table to configure the interrupt. So the DTS configuration will need to express priorities.

galak replied:
> We do set priorities. But it is not done in the IOAPIC. Priorities are reflected by the vector assignment in the IDT. Based on the priority level given to IRQ_CONNECT(), gen_idt uses this to find an appropriate range in the vector table to configure the interrupt. So the DTS configuration will need to express priorities.

Trying to understand this, based on include/arch/x86/arch.h - it looks like MVIC doesn't support priority, and ioapic is done by how vector values are assigned.  Is that correct?

ydamigos commented on dts/arc/quark_se_c1000_ss.dtsi:
```diff
@@ -61,5 +61,45 @@
+			compatible = "intel,qmsi-ss-i2c";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x80012000 0x400>;
```
Is 0x400 correct here?

tbursztyka replied:
good catch, seems like it's 0x100

ydamigos commented on arch/arc/soc/quark_se_c1000_ss/Kconfig.defconfig:
```diff
@@ -162,9 +162,6 @@ if I2C_QMSI
-config I2C_0_NAME
-	default "I2C_2"
-
 config I2C_0_IRQ_PRI
```
I am sure that I am missing something here. There is no interrupt property in intel,qmsi-ss-i2c but there is IRQ_PRI here.
Same for I2C_1.

ydamigos commented on arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series:
```diff
@@ -24,7 +24,7 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
```
Why not `if I2C_STM32`?

tbursztyka replied:
why? I did like it's done for SPI abve it.

ydamigos replied:
Sorry, I posted it in the wrong file.

ydamigos commented on dts/x86/quark_x1000.dtsi:
```diff
@@ -59,5 +59,17 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x90007000 0x400>;
+			interrupts = <18 IRQ_TYPE_LEVEL_LOW>;
```
Is IRQ_TYPE_LEVEL_LOW the IRQ priority?

tbursztyka replied:
No that's the interrupt flags. I don't know if it's specific to ioapic.

ydamigos reviewed (changes requested):
Some of my comments were hidden as outdated so I added them again.

ydamigos commented on arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series:
```diff
@@ -21,7 +21,7 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
```
Why not if I2C_STM32? The same for the other series.

tbursztyka replied:
Ah ok, I see.

ydamigos commented:
I could see a issue here. The second cell in interrupts property is used as IRQ priority in ARM device and as IRQ level in x86 (like in linux).

tbursztyka commented:
@galak, @ydamigos  could you re-review?

About the interrupt priority, it seems like an extract_dts_includes issue to me (among many others on that tool), so I won't address that in this PR.

tbursztyka commented:
@ydamigos , @galak I will make another PR for fixing this interrupt priority in intel's irq controllers, there are a lot of issues unrelated to i2c (uart not using CONFIG_ options for irq stuff, etc... it's a bit messy so it will require some cleanup patch first).

Commit:
dts/i2c: Add support for QMSI SS interrupt declaration in nodes

i2c qmsi ss driver requires 4 interrupts: error, rx, tx and stop.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Make QMSI SS driver IRQs set via CONFIG_ options

Now that dts i2 qmsi ss nodes generate the right options, let's use
them. Apply the relevant fixup on the targeted SoC.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tkln commented on dts/bindings/i2c/snps,designware-i2c.yaml:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: DesignWare SPI node description
```
Shouldn't this be I2C instead of SPI?

tbursztyka replied:
indeed, copy-pasted this from my other branch and forgot to change it

ydamigos commented on dts/bindings/i2c/st,stm32-i2c-v1.yaml:
```diff
@@ -27,12 +27,6 @@ properties:
       generation: define
       category: required
 
-    interrupts:
```
Why did you remove this? ...

tbursztyka replied:
good question...

ydamigos replied:
We shouldn't remove this.

ydamigos commented on dts/bindings/i2c/st,stm32-i2c-v1.yaml:
```diff
@@ -27,6 +27,12 @@ properties:
       generation: define
       category: required
 
+    interrupts:
```
And then you add it again.

tbursztyka commented:
just a pushed a new version: I was using node names such as i2c0@0x.... instead of i2c0@....

tbursztyka commented:
pushed again as master had an issue with ccs811

erwango reviewed (commented):
Looks good to me a part from last comment from @ydamigos (interrupt property removal in stm32 i2c v1 yaml binding)

ydamigos commented on drivers/i2c/Kconfig.stm32:
```diff
@@ -0,0 +1,48 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig I2C_STM32
```
STM32L0 series was added recently and it will be good to cleanup its Kconfig (arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series)

ydamigos reviewed (changes requested):
I would propose to add clock-frequency = <I2C_BITRATE_FAST>; for all boards in the PR because most users will use 400 kHz bitrate. If you don't, just keep in mind that every user should create a board overlay to override it in application level.

tbursztyka commented:
@ydamigos applied your comment. However, for the clock-frequency I applied what was existed previously i.e. the soc's dts (dts/<arch>/<soc>.dtsi is the on setting up the clock-frequency, and that seems to be always the standard speed. I don't see why I should change that, and if it has to: this change does not belong to this PR then. (it's not exactly the same subject).

ydamigos commented:
> However, for the clock-frequency I applied what was existed previously i.e. the soc's dts (dts//.dtsi is the on setting up the clock-frequency, and that seems to be always the standard speed. I don't see why I should change that, and if it has to: this change does not belong to this PR then. (it's not exactly the same subject).

I was talking about board's dts files. You add i2c nodes to them in this PR so you could add a line to override the default.

tbursztyka commented:
@ydamigos oh, yes, missed that.

galak commented:
Merged PR #6610 which was the same as this with minor cleanups.
